src/sentry/static/sentry/app/components/avatar.jsx

```diff
@@ -64,22 +94,19 @@ class Avatar extends React.Component {
       return null;
     }
     let user = this.props.user;
-    let avatarType = null;
-    if (user.avatar) {
-      avatarType = user.avatar.avatarType;
-    } else {
-      avatarType = user.email && this.props.gravatar ? 'gravatar' : 'letter_avatar';
-    }
+    let avatarType = this.getAvatarType();
     let props = {
       title: this.props.title,
       onError: this.onError,
       onLoad: this.onLoad,
     };
-    if (user.options && user.options.avatarType) {
-      avatarType = user.options.avatarType;
-    }
+
     if (avatarType === 'gravatar') {
-      return <img src={this.buildGravatarUrl()} {...props} />;
+      if (!this.state.gravatarUrl) {
```

Comment:
Part of me would prefer to isolate this deferred loading/rendering behavior to a dedicated `Gravatar` component. `Avatar` could just determine _which_ avatar type to load, and the sub components can abstract away the other details. Would also make testing cleaner.

Reply:
Agreed